labkey/server_context.py

```diff
@@ -81,23 +81,81 @@ def __init__(
     def __repr__(self):
         return f"<ServerContext [ {self._domain} | {self._context_path} | {self._container_path} ]>"
 
-    def build_url(self, controller: str, action: str, container_path: str = None) -> str:
-        sep = "/"
+    @property
+    def hostname(self) -> str:
+        return self._scheme + self._domain
 
-        url = self._scheme + self._domain
+    @property
+    def base_url(self) -> str:
+        base_url = self.hostname
 
         if self._context_path is not None:
-            url += sep + self._context_path
+            base_url += "/" + self._context_path
+
+        return base_url
+
+    def build_url(self, controller: str, action: str, container_path: str = None) -> str:
+        url = self.base_url
 
         if container_path is not None:
-            url += sep + container_path
+            url += "/" + container_path
         elif self._container_path is not None:
-            url += sep + self._container_path
+            url += "/" + self._container_path
 
-        url += sep + controller + "-" + action
+        url += "/" + controller + "-" + action
 
         return url
 
+    def webdav_path(self, container_path: str = None, file_name: str = None):
+        path = "/_webdav"
+        container_path = container_path or self._container_path
+
+        if container_path is not None:
+            if container_path.endswith("/"):
+                # trim the slash
+                container_path = container_path[0:-1]
+
+            if not container_path.startswith("/"):
+                path += "/"
+
+            path += container_path
+
+        path += "/@files"
+
+        if file_name is not None:
+            if not file_name.startswith("/"):
+                path += "/"
+
+            path += file_name
+
+        return path
+
+    def webdav_client(self, webdav_options: dict = None):
+        # We localize the import of webdav3 here so it is an optional dependency. Only users who want to use webdav will
+        # need to pip install webdavclient3
+        from webdav3.client import Client
```

Comment:
This is rad.

Reply:
Yeah it's rare that his feature is useful, but it's powerful when used!